Spectra-Fade (Invisibility potion)

TheShuEd opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->
Added a new Spectra-Fade reagent that can make any creature temporarily invisible. 

At the moment no methods of getting this reagent in the game have been added to PR, as I haven't played vanilla ss14 in a very long time, and to put it mildly I don't know where to put it to be balanced. Suggestions in the comments are welcome.

## Technical details
<!-- Summary of code changes for easier review. -->
StealthSystem is divided into several partial classes to separate the processing logic of StealthOnMove, the basic Stealth system, and the new TemporaryStealth module

Added new EntityEffect “AddStealth"

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->
![image](https://github.com/user-attachments/assets/37a4921c-3c0a-4941-a915-14c6b69c86a0)

https://github.com/user-attachments/assets/be5f8a5d-9ff4-416b-847f-70278d4e9c87

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

**Changelog**

:cl:
- add: TODO, when the in-game pickup method is added

ArtisticRoomba commented:
Holy shit

slarticodefast commented on Content.Shared/Stealth/Components/TemporaryStealthComponent.cs:
```diff
@@ -0,0 +1,28 @@
+    public float TargetVisibility = -1f;
+
+    [DataField, AutoNetworkedField]
+    public TimeSpan StartTime = TimeSpan.Zero;
```
needs timeoffsetserializer
also needs code comments for all the datafields

slarticodefast replied:
timeoffsetserializer is still missing 😞

TheShuEd replied:
timeoffsetserializer outdated, no? i add AutoPausedField and AutoGenerateComponentPause for handling pausing

slarticodefast replied:
no, it's still needed
without it it will serialize as the current server time

slarticodefast commented on Content.Server/EntityEffects/Effects/AddStealth.cs:
```diff
@@ -0,0 +1,31 @@
+        }
+
+        args.EntityManager.EntitySysManager.GetEntitySystem<StealthSystem>()
+            .AddTemporaryStealth(args.TargetEntity, TimeSpan.FromSeconds(amt));
```
Maybe it makes more sense to turn this into a status effect rather than using a custom entity effect

TheShuEd replied:
I'm not good with status effects, and I don't really like having to specify supported status effects on each creature type. 

I wouldn't want to convert this system to status effects.

slarticodefast reviewed (commented):
not a full review

Draedon4 commented:
the amount of ninja reports will skyrocket

TheShuEd commented:
Don't merge until it's fully added to the game
> Suggestions in the comments are welcome.

Callfromvoid commented:
What about make it hide only character, but not clothing and stuff in hands?

Entvari commented:
Would be pretty cool and fitting for the Wizard "uplink", maybe along with some other reagents.

muburu commented:
it'd be cool if it applied invisibility on skin contact so you could prank peeps by spraying them with invisible spray

TheShuEd commented:
Integrated into game:

- SpectraFade added into liquid anomaly fun reagent pool
- New SpectraFade ivisible reagent slime from anom supercrit
- Added into random maint pill reagent pool
- Added to Wizard store for 1 WizCoin

Ready to review and merg

TheShuEd commented:
https://github.com/user-attachments/assets/a5b57ecc-09e5-431b-927c-6d5490b8b9be

metalgearsloth commented on Content.Shared/Stealth/SharedStealthSystem.Move.cs:
```diff
@@ -0,0 +1,30 @@
+        SubscribeLocalEvent<StealthOnMoveComponent, MoveEvent>(OnMove);
+        SubscribeLocalEvent<StealthOnMoveComponent, GetVisibilityModifiersEvent>(OnGetMoveVisibilityModifiers);
+    }
+
+    private void OnMove(EntityUid uid, StealthOnMoveComponent component, ref MoveEvent args)
+    {
+        if (_timing.ApplyingState)
+            return;
+
+        if (args.NewPosition.EntityId != args.OldPosition.EntityId)
+            return;
+
+        var delta = component.MovementVisibilityRate * (args.NewPosition.Position - args.OldPosition.Position).Length();
+        ModifyVisibility(uid, delta);
+    }
```
These are really expensive are you sure you need this.

TheShuEd replied:
I didn't change it, I just moved it to a separate partial class

TheShuEd replied:
so, out of scope

EmoGarbage404 commented on Content.Server/EntityEffects/Effects/AddStealth.cs:
```diff
@@ -0,0 +1,31 @@
+    [DataField]
+    public float Time = 1;
+
+    protected override string? ReagentEffectGuidebookText(IPrototypeManager prototype, IEntitySystemManager entSys)
+        => Loc.GetString("reagent-effect-guidebook-add-stealth", ("chance", Probability));
```
you should integrate the time value into the guidebook text. You can probably just do something like "makes invisible for X seconds per 1u".

EmoGarbage404 commented on Content.Shared/Stealth/Components/TemporaryStealthComponent.cs:
```diff
@@ -0,0 +1,53 @@
+/// Some systems can make an item temporarily invisible.
+/// </summary>
+[RegisterComponent, NetworkedComponent, AutoGenerateComponentState, AutoGenerateComponentPause]
+public sealed partial class TemporaryStealthComponent : Component
```
missing Access on the component

EmoGarbage404 reviewed (changes requested):
Relatively small changes.

Also, i do kinda agree that this should be handled by status effects. Idk if it's really trivial to do right now since status effects are a little bit cooked though.

temm1ie commented:
> You definitely need to add this to the botany mutation pool for maximum botany gaming. Very please.

It would be possible to create a Kronkus plant with this chemical

TheShuEd commented:
This PR is waiting for #37238 now
